### ISSUE TITLE
File-order dependent checking reports Django mutable_list assignments

### ISSUE THREAD

**Issue by sudodaksh** [astral-sh/ty#4076]
## Summary

Repeated multithreaded checks of the same Django installation with the same ty binary produce different diagnostics in `django/contrib/gis/geos/mutable_list.py`. Restricting ty to one thread produces stable output. This was reproduced on an unchanged baseline binary.

## Environment

- ty: `ruff/0.15.22+65 (82005cc28 2026-07-21)`
- Python: 3.12
- Django: 6.0.7

## Reproduction

With a virtual environment containing Django 6.0.7:

```shell
SP=corpus-venv/lib/python3.12/site-packages
for i in 1 2 3; do
  ty check "$SP/django" --python corpus-venv --output-format concise 2>&1 | grep -c mutable_list.py
done
```

Repeated invocations of the same baseline binary produced 20, 22, and 23 diagnostics for `mutable_list.py`.

The intermittent diagnostics included:

```text
mutable_list.py:68:13: error[invalid-assignment] Object of type `bound method (Self@__init__ & ~<Protocol with members _set_single>)._set_single_rebuild(...) -> Unknown` is not assignable to attribute `_set_single` on type `Self@__init__ & ~<Protocol with members _set_single>`
mutable_list.py:69:13: error[invalid-assignment] Object of type `bound method (Self@__init__ & ~<Protocol with members _set_single>)._assign_extended_slice_rebuild(...) -> Unknown` is not assignable to attribute `_assign_extended_slice` on type `Self@__init__ & ~<Protocol with members _set_single>`
```

A single-threaded run was stable:

```shell
TY_MAX_PARALLELISM=1 ty check "$SP/django" --python corpus-venv --output-format concise
```

It reported 1,857 diagnostics overall on repeated baseline comparisons, without the intermittent assignments.

## Expected behavior

Diagnostics should be deterministic across runs and independent of thread scheduling.

**Comment by MichaReiser** [comment]
Minimal reproducer according to codex:

`base.py`:

```py
class Base:
    def __init__(self):
        if not hasattr(self, "x"):
            self.x = self.__str__
```

`child.py`:

```py
from base import Base


class Child(Base):
    x = Base.__str__
```

[Playground](https://play.ty.dev/a9fa5f9c-56ad-4089-9d3e-f07b8d76b5e3)

Checking base first fails

```bash
$ TY_MAX_PARALLELISM=1 ty check base.py child.py --output-format concise
base.py:4:13: error[invalid-assignment] Object of type `bound method (Self@__init__ & ~<Protocol with members 'x'>).__str__() -> str` is not assignable to attribute `x` on type `Self@__init__ & ~<Protocol with members 'x'>`
Found 1 diagnostic
```

Checking `child` first succeeds

```bash
$ TY_MAX_PARALLELISM=1 ty check child.py base.py --output-format concise
All checks passed!
```

**Comment by charliermarsh** [comment]
It's looking like this is ultimately a case of `Self` with an intersection type and could be solved by https://github.com/astral-sh/ty/issues/3731, https://github.com/astral-sh/ty/issues/2455, etc. (As in, one side of the non-deterministic results leads to an error due those issues, while the other succeeds.) But, there might be some separate determinism fix here too.

**Comment by dcreager** [comment]
If the non-determinism is coming from the constraint set implementation, the new ["wobble tests"](https://github.com/astral-sh/ruff/pull/26896) are a good way to exercise it. We've been putting nondet reproducers into the `constraint_set_ordering.md` mdtest file, where running it with different values of `TY_CONSTRAINT_SET_ORDER` produces different results.

**Comment by charliermarsh** [comment]
Thanks. I think this is a bit different. The non-determinism is introduced based on the order in which files are checked, which gives us a different Salsa cycle head.

**Comment by carljm** [comment]
Fixing the rule that triggers a diagnostic could paper over this specific instance, but it would be great if we can actually fix the non-determinism. Which probably involves figuring out why the fixpoint iteration resolves differently with a different cycle head.

**Comment by charliermarsh** [comment]
Yup, that’s the plan!

**Comment by charliermarsh** [comment]
Checkpointing my understanding... Given:

```python
# base.py
class Base:
    def __init__(self):
        if not hasattr(self, "x"):
            self.x = self.__str__

# child.py
from base import Base

class Child(Base):
    x = Base.__str__
```

Results currently vary depending on whether you check `base.py` or `child.py` first.

IIUC there are three things we need to compute when checking:

1. Narrow `self` to `Self & ~Protocol[x]`
2. Infer the type of `Base.x`
3. Infer the right-hand side of `self.x = self.__str__`
 
If you check `base.py` first, we begin inside `Base.__init__`:

- We narrow self to `Self & ~Protocol[x]` (1)...
- Constructing `Self & ~Protocol[x]` requires determining whether `Self` already has `x`...
- Determining whether `Self` already has `x`, means we have to infer `Base.x` (2)...
- Which means we need its initializer for `self.x = self.__str__` (3)...
- Which means we need the narrowed receiver: `Self & ~Protocol[x]` (1)

So `Self & ~Protocol[x]` (1) is the cycle head.

When checking `child.py` first, we try to validate the subclasses overrides:

- We need to infer `Child.x`
- We need to infer `Base.x` (2)
- ...

So inferring `Base.x` (2) is the cycle head.

When (1) is the cycle head, it's seeded with `Type::divergent(id)`.  But when (2) is the cycle head, it's seeded with `Member { inner: Place::bound(Type::divergent(id)).into() }`, which means that the attribute _exists_ but the type is `Divergent`. And if `x` exists, `not hasattr(self, "x")` is unreachable, so the whole block is suppressed.

As an example, if you check this instead:

```python
from typing import reveal_type

class Base:
    def __init__(self):
        if not hasattr(self, "x"):
            reveal_type(self)
            reveal_type(self.__str__)
            self.x = self.__str__
            reveal_type(self.x)
```

Then if you check `base.py` first, you get:
```
Revealed type: Self@__init__ & ~<Protocol with members 'x'>
Revealed type: bound method (Self@__init__ & ~<Protocol with members 'x'>).__str__() -> str
error[invalid-assignment]
Revealed type: bound method Self@__init__.__str__() -> str
```

If you check `child.py` first, you get:
```
Revealed type: Never
Revealed type: Never
Revealed type: Never
```

**Comment by carljm** [comment]
Yeah, it seems like the core mismatch here is that we assume attributes are always definitely present on a type which says they are present, and we narrow / constrain reachability from `hasattr` checks accordingly -- but we allow attributes to be conditionally defined in `__init__.py`, and respect reachability there. A narrowly-targeted fix might be to never deduce definite reachability from a `hasattr` check on `self`, in any method? That is, make a partial allowance for the possibility that methods of a class might see partially-initialized instances. (We could also target this even more narrowly to `__init__`, though it could potentially happen in other methods too, since we allow attributes to be defined in any method.)

**Comment by charliermarsh** [comment]
(I think it needs to be a bit broader than that to also take into account cases like https://github.com/astral-sh/ty/issues/4221.)

**Comment by MichaReiser** [comment]
Nice work @charliermarsh !
